command/agent/command.go

```diff
@@ -547,13 +547,49 @@ func (c *Command) Run(args []string) int {
 	// Start retry join process
 	c.retryJoinErrCh = make(chan struct{})
 
-	joiner := retryJoiner{
-		join:     c.agent.server.Join,
-		discover: &discover.Discover{},
-		errCh:    c.retryJoinErrCh,
-		logger:   c.agent.logger,
+	if config.Server.Enabled && len(config.Server.RetryJoin) != 0 {
```

Comment:
Make sure there is verification that:

1. On the server either, config.Server.RetryJoin/StartJoin/etc is set or the new ServerJoin is set and not both
2. On the client that StartupJoin isn't set.